### ISSUE TITLE
Julia crashes when doing levels! on a DataFrame column

### ISSUE THREAD

**Issue by pascalgolec** [JuliaData/DataFrames.jl#1634]
Julia 1.0.2, DataFrames v0.15.2 and CategoricalArrays v0.5.2 gives me an interpreter error that crashes julia when using levels! on a DataFame column directly:

```julia
# crashes:
using DataFrames
mycats = ["a", "b"]
mydf = DataFrame(x = ["a", "b", "a"])
function cat!(df::DataFrame, cats::Vector{String})
    df.x = CategoricalArray(df.x)
    levels!(df.x, cats) # julia crashes here
    return df
end
mydf = cat!(mydf, mycats)
```

These two scenarios on the other hand work:

```julia
# works if levels! is outside of function
using DataFrames
mycats = ["a", "b"]
mydf = DataFrame(x = ["a", "b", "a"])
function cat!(df::DataFrame, cats::Vector{String})
    df.x = CategoricalArray(df.x)
    return df
end
mydf = cat!(mydf, mycats)
levels!(mydf.x, mycats)
```

```julia
# works if don't apply levels! on a DataFrame column:
using CategoricalArrays
mycats = ["a", "b"]
x = ["a", "b", "a"]
function cat!(vec::Vector{String}, cats::Vector{String})
    vec = CategoricalArray(vec)
    levels!(vec, cats) # julia crashes here
    return vec
end
cat!(x, mycats)
```

**Comment by bkamins** [comment]
I think this is a known issue with Julia compiler. CategoricalArrays.jl stress it to the extreme. Similar issues have been reported, see https://github.com/JuliaLang/julia/issues/25430, but I think @nalimilan might know if they cover this case.

**Comment by nalimilan** [comment]
Yes I think that's https://github.com/JuliaLang/julia/issues/29430. It's fixed in Julia 1.1. Unfortunately it hasn't been backported to 1.0.3 because it could break existing code, and we haven't been able to run checks on packages to see whether that's the case or not.
